### ISSUE TITLE
LCE seclector issue

### ISSUE THREAD

**Issue by oshtaier** [SatelliteQE/airgun#51]
Here is issue that I faced and cannot resolve in any way that I tried. Already spent too much time, so fresh look is needed. 
Seems as architectural blocker.
Quick reproducer:
```
ak_name = gen_string('alpha')
    non_searchable_ak_name = gen_string('alpha')
    org = entities.Organization().create()
    envs_list = ['STAGING', 'DEV', 'IT', 'UAT', 'PROD']
    for name in envs_list:
        entities.LifecycleEnvironment(name=name, organization=org).create()
    env_name = random.choice(envs_list)
    cv = entities.ContentView(organization=org).create()
    cv.publish()
    promote(
        cv.read().version[0],
        entities.LifecycleEnvironment(name=env_name).search()[0].id
    )
    with session:
        session.organization.select(org_name=org.name)
        for name in [ak_name, non_searchable_ak_name]:
            session.activationkey.create({
                'name': name,
                'lce': {env_name: True},
                'content_view': cv.name
            })
            assert session.activationkey.search(name) == name
```
so you ends up in:
![image](https://user-images.githubusercontent.com/10895065/37969654-f565b860-31d9-11e8-912c-cbcf7d7a3d0d.png)

Here is way of my thoughts, but it is obviously wrong:
```
class LCESelectorGroup(Widget):
    ROOT = ".//div[@path-selector='environments']"
    ITEMS = "./div"

    def get_selector(self, criteria):
        """Return all lce selectors from the group"""
        # return [
        #     LCESelector(selector.locator)
        #     for selector
        #     in self.browser.elements(self.ITEMS)
        # ]
        # or
        # for selector in self.browser.elements(self.ITEMS):
        #   if criteria in selector.get_attribute('innerHTML'):
        #       return LCESelector(selector.locator)

    def fill(self, value):
        checkbox_name = list(value.keys())[0]
        selector = self.get_selector(checkbox_name)
        return selector.fill(value)
```
or through
```
    ROOT = ".//div[@path-selector='environments']"
    item = LCESelector("./div[contains(., '{}')]")
```

1) You cannot work with `LCESelector` or any widget outside of initialization as you finish up with `WidgetDescriptor`. Anyway it will not help a lot as implementation will look really bad
2) You cannot initialize widget on view level as you don't know name of LCE in advance
3) Working with names anyway seems as not the best option, but till that moment seems as only one possible. If we go with names it will be impossible to select 'Library' checkbox in third row for example, but if not names, there are nothing unique at all. Indexes are obviously not acceptable idea.
4) It is not nested view or view at all. Basically, I don't like idea of lce selectors group in widget format too, but we cannot modify `LCESelector` in any way as it should not have any connection to group or parent frame as it is wrong approach from architecture point of view

**Comment by abalakh** [comment]
Two options off the top of my head - either using locator + browser.element directly or using [Parametrized Views](https://up.falesnik.net/wt-doc/advanced_usage.html#parametrized-views). That's in case we want to preserve original widget. But should we? Is there a place where only 1 lce is shown, not all the available for current organization?

**Comment by oshtaier** [comment]
I am interested only in technical side of implementation. Okay, so you proposing:
```
session.activationkey.create(
            'env_name': 'test',
        {
            'name': name,
            'lce': {env_name: True},
            'content_view': cv_name
        })
```
```
    def create(self, values, env_name):
        view = self.navigate_to(self, 'New', env_name=env_name)
        view.fill(values)
        view.submit.click()

then put here
@navigator.register(ActivationKeyEntity, 'New')
class AddNewActivationKey(NavigateStep):
```
```
class ActivationKeyCreateView(BaseLoggedInView, ParametrizedView):
```
as you see that is really bad idea

**Comment by oshtaier** [comment]
or you propose 
`class LCESelectorGroup(ParametrizedView):`
?

**Comment by oshtaier** [comment]
Anyway, please provide complete solution

**Comment by abalakh** [comment]
None of them. Sorry, forgot to mention - take a look at second example from that link, it illustrates the idea. There should be parametrized view inside ActivationKeyCreateView, which accepts lce name, smth similar to that example:
```python
def fill(self, values):
    ...
    self.lce_group('foo').lce_widget.fill()
```
At least that's the direction i've been told to look into when asked about pretty similar case.

**Comment by oshtaier** [comment]
not the worst idea and maybe our only one possible shot, but with that approach we still never be able to click for example Library from 3 row or something non unique

**Comment by abalakh** [comment]
Doesn't seem like we have many options, considering there's no way to identify specific lceselector in group other than lce name.
But i have 1 idea. We can keep similar interface (`"lce_name": True`) but extend it for such corner cases when smth non-unique is needed: instead of "True" we may allow to specify concrete lce from group containing that "lce_name". E.g., for your third group: `'PROD':  'Library'`, this way we'll identify desired group by name "PROD" and select "Library" in it. And that's corner case which will be used rarely, most common one will remain the same - you'll be able to specify just `'PROD': True` to select "PROD".
@oshtaier thoughts? I know it's not perfect solution, but that's definitely not perfect UI which doesn't give us plenty of alternatives.

**Comment by oshtaier** [comment]
hmm, doesn't seem that we need to make logic more complex till we meet situation like that, let's stick to initial plan. Fine to implement the fix?

**Comment by oshtaier** [comment]
good luck

**Comment by abalakh** [comment]
@oshtaier progress so far:
LCE group looks like that:
```python
class LCESelectorGroup(ParametrizedView):
    ROOT = ".//div[@path-selector='environments']"

    PARAMETERS = ('lce_name',)

    lce = LCESelector(
        locator=(
            "//div[contains(@class, 'path-selector')]"
            "[ul/li[normalize-space(.)='{lce_name}']]")
    )

    def fill(self, values=None):
        if values is True:
            values = {self.context['lce_name']: True}
        else:
            values = {values: True}
        return self.lce.fill(values)
```
It connects to view pretty easily:
```python
class ActivationKeyCreateView(BaseLoggedInView):

    name = TextInput(id='name')
    ...
    lce = ParametrizedView.nested(LCESelectorGroup)
    ...
    submit = Text("//button[contains(@ng-click, 'handleSave')]")
```
(can be added as an included class as well, but turns out there's more elegant 1-linear way to do that)

From test perspective interface is literally the same, thanks to custom `.fill()` implementation:
```python
    session.activationkey.create({
            'name': name,
            'lce': {env_name: True},
            'content_view': cv_name
        })
```

Question - is this implementation ok to you, should i continue or you have some concerns/other ideas?
Those ParametrizedViews are a bit tricky and i still need to figure out how to make `.read()` work with it (custom `all()` implementation, i guess), which will turn into waste of a day or two if parametrizedviews are no go for us :)

**Comment by oshtaier** [comment]
Is there reason for such `fill` implementation? It looks a little bit weird. And what is a way to pass `False` there?

**Comment by abalakh** [comment]
@oshtaier it's basically a shortcut to pass info to included widget (lceselector), so in the end we can use it like `'lce': {env_name: True},`. Without custom .fill() we would have to pass smth like `'lce': {'lce': {env_name: True}}`.
Regarding `False` - should it be supported? I thought those are 'fake' checkboxes which are rather radio buttons, so to deselect one lce we should select another one. But if i got it wrong - no issues, I can change first 2 lines of `.fill()` to:
```python
    if values in (True, False):
            values = {self.context['lce_name']: values}
```
and it will work.
`else` block adds ability of selecting other lce in the same selector, like i've previously mentioned with 'Library'. It was very easy to support that with current logic, just 1 extra line. I can remove it if you insist, but i'd rather like to have it.

**Comment by oshtaier** [comment]
yes, please add support for False and `else` is not bad idea, but require very detailed docstring as it is hard to understand that logic

**Comment by abalakh** [comment]
@oshtaier There's another aspect we should consider. With ParametrizedViews approach, we should always have parameter (lce name, in our case) which represents single lce selector which is used in locator to identify that specific widget. Currently i'm using last lce name available in lceselector as such id. This means `view.read()` will return things differently, example for your screenshot:
```python
{
    'name': 'some_name',
    'description': 'some_description',
    'lce': {
        'DEV': {'Library': False, 'DEV': False},
        'IT': {'Library': False, 'IT': False},
        'PROD': {'Library': False, 'PROD': False},
        [...]
    },
}
```
On the one hand, that extra dict is needed, on the other - it makes comparing values in test layer a bit tricky.
When we know for sure last name in lce selector we can compare things like
```python
assert ak_values['lce'][env_name][env_name] == True
```
If we don't know specific name or wanna make sure some env is present somewhere in any of selectors it becomes smth like
```python
assert any(lce[env_name] == True for lce in ak_values['lce'].values())
```
Suggestions, objections?

**Comment by oshtaier** [comment]
I think that we should stop on idea to support selection/deselection only latest element in the row to not make solution too complex.
And provide direct URL to that issue in docstring in case someone will ever need ability to select any checkbox in the selector which should not happen anyway.

**Comment by abalakh** [comment]
@oshtaier that won't change a thing, extra dict is introduced by parametrizedview, not by ability to select specific lce. I was talking about `.read()`, not `.fill()`
